lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+                                                  context.getAllocator())});
+  graphConfiguration->handleDevicePlacement(
+      "Const",
+      /*opDevice*/ getDeviceString(DeviceType::ALL), builder, location,
+      attributes);
+  GraphOperationInst *constNode = builder.createGraphOperation(
+      location, context.getIdentifier(opName), /*operands*/ {}, attributes,
+      {convertElementTypeToTensorValueType(intType)});
+  assert(constNode->getNumResults() == 1);
+  return constNode->getResults()[0];
+}
+
+std::pair<SILBasicBlock *, SILValue>
+SingleExitLoopTransformer::createNewHeader() {
+
+  SILBuilder builder(header);
+  ASTContext &context = builder.getASTContext();
+
+  SILBasicBlock *newHeader = currentFn->createBasicBlock();
+  loop->addBasicBlockToLoop(newHeader, LI->getBase()); // Should be done first.
+  // Clone arguments and change all uses to the new header's arguments.
+  newHeader->cloneArgumentList(header);
+  SmallVector<SILValue, 8> headerArgs;
+  for (auto i : indices(header->getArguments())) {
+    header->getArgument(i)->replaceAllUsesWith(newHeader->getArgument(i));
+    headerArgs.push_back(newHeader->getArgument(i));
+  }
+  header->dropAllArguments();
+  // Add phi arguments in the new header corresponding to the escaping values.
+  for (const auto &escapingValue : escapingValues) {
+    SILValue newValue = newHeader->createPHIArgument(
+        escapingValue->getType(), escapingValue.getOwnershipKind());
+    escapingValue->replaceAllUsesWith(newValue);
+  }
+  // An integer to identify the exit edge.
+  SILValue exitIndexArg = newHeader->createPHIArgument(
+      convertElementTypeToTensorValueType(
+          SILType::getBuiltinIntegerType(32, context)),
+      ValueOwnershipKind::Owned);
+  // A boolean corresponding to the stayInLoop flag.
+  newHeader->createPHIArgument(convertElementTypeToTensorValueType(
+                                   SILType::getBuiltinIntegerType(1, context)),
+                               ValueOwnershipKind::Trivial);
+  return std::make_pair(newHeader, exitIndexArg);
+}
+
+SILBasicBlock *
+SingleExitLoopTransformer::createLatchBlock(SILBasicBlock *newHeader) {
+  // Create a new latch block.
+  SILBasicBlock *latchBlock = currentFn->createBasicBlock();
+  latchBlock->cloneArgumentList(newHeader);
+  SILBuilder  builder(latchBlock);
+  SmallVector<SILValue, 8> latchArgs;
+  for (const SILArgument *latchArg : latchBlock->getArguments()) {
+    latchArgs.push_back(latchArg);
+  }
+  builder.createBranch(
+    getUserSourceLocation(header->getTerminator()->getDebugLocation()),
+      newHeader, latchArgs);
+  loop->addBasicBlockToLoop(latchBlock, LI->getBase());
+  return latchBlock;
+}
+
+llvm::DenseMap<SILBasicBlock *, intmax_t>
+SingleExitLoopTransformer::patchEdges(SILBasicBlock *newHeader,
+                                      SILBasicBlock *latchBlock) {
+
+  llvm::DenseMap<SILBasicBlock *, intmax_t> exitIndices;
+
+  size_t oldHeaderNumArgs =
```

Comment:
should we save oldHeaderNumArgs at initialize() time, and just do an assert for sanity check here?

Reply:
it just feels round-about that we are recovering / reverse-engineering some info that we could get from the input. instead we are mutating the input state, and then cancelling out its effect to get that same info.